### PR TITLE
CY-730 : Add tracing framework

### PR DESCRIPTION
**Pull request by davidginzbourg**
The following features work whenever `enable_tracing` is set to `true` and
`tracing_endpoint_ip` is also set with the corresponding Jaeger API Server
IP.
- Adds automatic tracing to thread context whenever an API call is made
- Adds a decorator for tracing, it starts a new span before the function
is called.

### CODE REVIEW

File: rest-service/manager_rest/config.py [hunk 1]
```diff
@@ -79,10 +80,15 @@ def __init__(self):
...
 
+        self.enable_tracing = False
+        self.tracing_endpoint_ip = None
+        self.tracer_config = None
+
     def load_configuration(self):
         for env_var_name, namespace in CONFIG_TYPES:
             if env_var_name in os.environ:
                 self.load_from_file(os.environ[env_var_name], namespace)
+        self._create_tracer_config()
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/config.py]: It does load the attributes from the main config file. Afterwards it goes to creates a `jaeger_client.Config()` instance. The instantiation used to be in the Flask application but been moved by Tal's request to `config.py`.

File: rest-service/manager_rest/config.py [hunk 2]
```diff
@@ -97,6 +103,18 @@ def load_from_file(self, filename, namespace=''):
                     "Ignoring unknown key '{0}' in configuration file "
                     "'{1}'".format(key, filename))
 
+    def _create_tracer_config(self):
+        if not self.enable_tracing or not self.tracing_endpoint_ip:
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/config.py]: The loop in `load_configuration`. These attributes are loaded from the yaml file which resides at `MANAGER_REST_CONFIG_PATH`

File: rest-service/manager_rest/rest/rest_decorators.py [hunk 3]
```diff
@@ -40,6 +40,7 @@
     Schema,
 )
 from ..security.authentication import authenticator
+from manager_rest.utils import traces
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/rest/rest_decorators.py]: Done

File: rest-service/manager_rest/security/authorization.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 from flask import request
 from flask_security import current_user
 
+from manager_rest.utils import traces
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/security/authorization.py]: Done

File: rest-service/manager_rest/server.py [hunk 5]
```diff
@@ -149,6 +172,33 @@ def handle_exception(flask_method, flask_restful_method, e):
...
+        This function creates a new span and injects it to the current thread
+        local for other inner executions to fetch.
+        """
+        if not self.tracer:
+            return super(CloudifyFlaskApp, self).dispatch_request()
+
+        request = stack.top.request
+        operation_name = '{} ({})'.format(request.endpoint, request.method)
+        headers = {}
+        for k, v in request.headers:
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/server.py]: It fetches the span context if it appears in the headers

File: tests/integration_tests/tests/test_cases.py [hunk 6]
```diff
@@ -470,6 +471,16 @@ def client_using_tenant(self, client, tenant_name):
         finally:
             client._client.headers[CLOUDIFY_TENANT_HEADER] = curr_tenant
 
+    def _setup_tracing(self):
+        """Updates the rest server config to enable tracing if
+        TRACING_ENDPOINT_IP env var is set.
+        """
+        tracing_endpoint_ip = os.environ.get('TRACING_ENDPOINT_IP', None)
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply tests/integration_tests/tests/test_cases.py]: Done

File: rest-service/manager_rest/config.py [hunk 7]
```diff
@@ -97,6 +103,18 @@ def load_from_file(self, filename, namespace=''):
...
+            return
+        self.tracer_config = jaeger_client.Config(
```
[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/config.py]: @tyacbovi ?

[the remark this answers is not in the document]
Comment (davidginzbourg) [reply rest-service/manager_rest/config.py]: Done.

**Comment by geokala** [comment]
This has been open for more than six months so I'm closing it. If it's still wanted, please reopen it and ping me so we can get it reviewed.

**Pull request closed without merging**